README.md
# moveVis
[![CRAN version](https://www.r-pkg.org/badges/version/moveVis)](https://CRAN.R-project.org/package=moveVis)
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/last-month/moveVis?color=brightgreen)](https://CRAN.R-project.org/package=moveVis)

## Introduction

This is an R package providing tools to visualize movement data by creating path animations from geo-location point data. The package is under ongoing development.  The moveVis package is working hand in hand with the move package by using the move and moveStack class and the raster package. It is based on a ggplot2 plotting architecture and relys on the libraries ImageMagick, ffmpeg and libav. To be informed about updates, new features and the current version, visit [news.movevis.org](http://news.movevis.org).

## Examples

Output of animate_move(), showing White Storks movement nearby Lake Constance, using a static land cover/land use map in the background:

![Alt Text](https://github.com/16EAGLE/AUX_data/blob/master/data/examp1.gif)

Output of animate_move(), showing White Storks movement nearby Lake Constance, using a dynamic MODIS NDVI layer in the background:

![Alt Text](https://github.com/16EAGLE/AUX_data/blob/master/data/examp2.gif)

## Installation

This is the official moveVis R package repository, including beta code versions before submitted to CRAN. For operational use of moveVis, please use the current stable CRAN version of moveVis.

To install stable version from CRAN, please execute:

```s
install.packages('moveVis')
```

To install the development version from this GitHub repository, please execute:

```s
devtools::install_github("16EAGLE/moveVis")
```

## Getting started

You can use moveVis with any move or moveStack object. This guide shortly explains how to prepare your own geo-location point data for the animate_move() function by creating a move class object from a data.frame class object. As an example, the provided example data (data.frame) are used. Instead, you could use any similar prepared data of yours.

First, load moveVis and the move package:


```s
#Load packages
library(moveVis)
library(move)
```

### External libraries

moveVis requires at least one of the three external libraries 'ffmpeg', 'libav' and/or 'ImageMagick'. They support different types of output formats (gif, mov, mp4 etc.). If you have 'ImageMagick' and either 'ffmepg' or 'libav' installed, you can use all output formats supported by moveVis.

Run `get_libraries()` to find out, which libraries are installed on your system and to get instructions how to download and to install the needed libraries:

```s
get_libraries()
```

#### ImageMagick

For GIF outputs, ImageMagick is needed. On many Linux distributions, ImageMagick is preinstalled or can be installed via the package manager (on Ubuntu: `sudo apt-get install imagemagick`). On Windows, Install ImageMagick from https://www.imagemagick.org/script/download.php. Make sure that you select "Install legacy utilities (e.g. convert)" during the installation process. 

#### ffmpeg/libav

To be able to output video formats, ffmpeg is recommended. Ubuntu users can easily install it executing `sudo apt-get install ffmpeg` from the terminal (similar on other Linux platforms). On Windows, installation takes a little more steps, which are nicely described in this recommended guide (including the link to the binary download, which you need): https://video.stackexchange.com/questions/20495/how-do-i-set-up-and-use-ffmpeg-in-windows/20496#20496. Alternatively, moveVis supports libav, which is preinstalled on some Linux distributions.

After the required libraries are once installed, <b>restart your R session</b> and then run `get_libraries()` again to check, if they are recognized by moveVis. If so, everything is set for starting to create your first moveVis animation.

### Data preperation

You will need to load the example data for this tutorial:

```s
#Load data (data.frame) (or use your own as data.frame)
data("move_data")
```

As the provided example data, your data.frame needs to have at least three columns: two columns for your coordinates (here "lat", "lon") and one for the date/time stamp (here "dt"). The date/time stamps need to be converted to POSIXct as follows:

```s
move_data$dt <- as.POSIXct(strptime(move_data$dt, "\%Y-\%m-\%d \%H:\%M:\%S", tz = "UTC"))
```

Your movement data need to be provided as move class objects to the animate_move() function. For each individual movement path you want to display simultaniously within a single animation, you will need one move class object. The move class objects per path are provided as a list. If your data.frame contains several individuals (e. g. differentiable by a "individuals" column, as the example data.frame does), then subset the data per individual and store the namings. If you just want to display a single path, you do not have to do this.

```s
#Create new move class object list by individual
data_ani <- split(move(move_data$lon, move_data$lat, proj=CRS("+proj=longlat +ellps=WGS84"),
                       time = move_data$dt, animal=move_data$individual, data=move_data))
```

### Additional input variables

get_libraries() returns the library commands that are needed by the animate functions. Just save them to a variable that you can later pass to the animate function so that it knows how to call the extern library commands. You can also call get_formats() to see all output formats, you can choose from.


```s
#Get libraries 
conv_dir <- get_libraries()

#Find out, which output file formats can be used
get_formats()

```

Last, you need to specify the output directory path and you can specify some optional variables such as the animation title (for details on all the arguments of animate_move() , read the animate_move() help).

```s
#Specify output directory
out_dir <- "/your/full/output/directory"

#Specify some optional appearance variables
img_title <- "Movement of the white stork population at Lake Constance, Germany"
img_sub <- paste0("including individuals ",indi_names)
img_caption <- "Projection: Geographical, WGS84; Sources: Movebank 2013; Google Maps"
```

### Animate function call

Finally, you are now prepared to call animate_move(), which will have to work for a while depending on your input. Here, for demonstrational purposes, we use `frames_nmax` set to 50 to force the function to only produce 50 frames and then finish the animation, regardless how many input points you provided. Set `log_level` to 1 to be informed of anything the function is doing. Set `out_format` to "mov" to get a .mov video output file.

```s
#Call animate_move()
animate_move(data_ani, out_dir, conv_dir = conv_dir, tail_elements = 10,
             paths_mode = "true_data", frames_nmax = 50,
             img_caption = img_caption, img_title = img_title,
             img_sub = img_sub, log_level = 1, out_format = "mov")
```

After the function is finished, check the output directory. Retry everything with different settings and modes, described in the function manuals. Further examples and explanations are provided within the function manuals.

## Contact & bug reports

moveVis is being developed and maintained by Jakob Schwalb-Willmann. For bug reports, please use <https://github.com/16eagle/movevis/issues> to contact me. Feature requests and other contributions are also welcome.

## What else are we doing?

The Department of Remote Sensing of the University of Würzburg has developed other R packages that might interest you:
 * <a target="_blank" href="http://bleutner.github.io/RStoolbox/">RStoolbox</a>
 * <a target="_blank" href="https://github.com/RRemelgado/rsMove/">rsMove</a>

For other news on the work at at the Department of Remote Sensing of the University of Würzburg, click <a target="_blank" href="http://remote-sensing.eu/">here</a>.


## Acknowledgements
          
This initiative is part of the <a target="_blank" href="http://www.fernerkundung.geographie.uni-wuerzburg.de/forschung/projekte/laufende_projekte/opt4environment">Opt4Environment</a> project and was funded by the German Aerospace Center (DLR) on behalf of the Federal Ministry for Economic Affairs and Energy (BMWi) with the research grant <b>50 EE 1403</b>.

<p align="justify">
<a href="http://www.fernerkundung.geographie.uni-wuerzburg.de/en/lehrstuehle_und_arbeitsgruppen/department_of_remote_sensing/startseite//"><img width="150" height="100" src="https://www.uni-wuerzburg.de/typo3conf/ext/uw_sitepackage/Resources/Public/Images/uni-wuerzburg-logo.svg"></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="http://www.dlr.de/eoc/en/"><img width="115" height="100" src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f5/DLR_Logo.svg/744px-DLR_Logo.svg.png"></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="http://www.bmub.bund.de/"><img width="220" height="100" src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRX92Q6lhYFo0Rv7p7Y3obqFXsxRyjXMNKSJ_q9bAvXYdFd5wOF3Q"></a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="http://www.orn.mpg.de/en/"><img width="200" height="100" src="https://www.molgen.mpg.de/188611/mpi_Seew_LogoText-1355515314.gif"></a>

</p>


NEWS
----------------------------------------------------------
moveVis 0.9.7 (currently working on, no release so far)

----------------------------------------------------------- 
 
To-Do:
 + "keep tracks" setting to force paths to not disappeare
 + segmentation coulouring [working on this at the moment]
 + optional user-defined font sizes of all fonts (axis, annotations etc.), colours based on ggplot2 syntax
 + optional user-defined map elements position vectors
 + user-defined area of interest (AOI), which scales map extent to specific area without excluding tracks "out of sight", enables move-in and move-out of individuals to AOI
 + follow population mode
 + follow individual mode
 + day/night-mode visualization
 + dynamic overplotting, e. g. polygons or points that change over time

----------------------------------------------------------
moveVis 0.9.6 (stable release, latest on CRAN)
automatic preprocessing, video support, better performance, bug fixes
-----------------------------------------------------------
New functions:
 + get_libraries() handles all extern libraries that are needed for video formats. Just run it to get information on what you need to install. Replaces get_imconvert(), which can still be called for code compatibility reasons
 + get_formats() outputs all file formats that can be used with moveVis depending on which libraries are installed on the system.

New features:
 + moveVis supports now multiple video formats in addition to GIF, if it is linked to a video library (either FFmpeg of libav)
 + animate_move() now detects temporal resolution and uniformity of timestamps automatically to determine, if interpolation needs to be applied to calculate uniform frame times per frame tick
 + animate_move() now calculates and displays the final animation duration derived from the total number of output frames and the fps prior to generating each frame, so that the user can already approximate which size the output animation will have.
 
 + added "frames_tres" to animate_move() to change temporal resolution through linear interpolation
 + added "frames_pixres" to animate_move() to adjust frame ppi
 + added "paths_na.hold" to animate_move() for defining how to deal with data gaps (hold or not hold last path location)
 + added "time_bar_col" to animate_move() for changing colouring of the time bar at the top border of the map
 + added "out_format" to animate_move() for defining output file format
 + added "indi_names" to animate_move() for defining individual names vector manually
 + added "scalebar_dist" to animate_move() for defining the scalebar length manually
 + added "overwrite" to animate_move() to regulate output file writing behaviour
 
Improvements:
 + it is now possible to use the 'layer' argument to specify one (static) raster object without specifying 'layer_dt', which is only needed, if several raster objects are handed over as a list.
 + in some situations, the movement data extent was bigger than the cropped layer extent: this works now
 + in some situations, stat animation lines were cut off, this should be resolved now
 + recoding of multiple processing steps to increase processing speed and stability
 + increased error handler by several errors likely to occure
 + several bug fixes

Contributors:
 + AniMove 2017 participants
 
----------------------------------------------------------
moveVis 0.9.5
adding frames_layout, static_data etc., improvements to workflow
-----------------------------------------------------------
New features:
 + RGB stats support added for animate_stats() and animate_move()
 + added animate_move() argument "time_scale" to enable a time scale
 + added animate_move() argument "frame_layout" for user friendly, complex layouting of the output GIF by specifiying the plot locations
 + added animate_move() argument "layer_stretch"
 + added animate_move() expert arguments "conv_cmd" and "conv_frames"

Improvements:
 + several major bug fixes
 + major animate_raster() bug fixes, reenabling use of RGB rasterStack inputs
 + stats legend is now locatable using frame_layout
 + fixed a bug of get_imconvert() that caused it to have issues with finding convert.exe on Windows machines, if ImageMagick is preinstalled

Contributors:
 + Bart Kranstauber

----------------------------------------------------------
moveVis 0.9.4
adding animate_stats() and stats arguments for animate_move()
-----------------------------------------------------------
New features:
 + animate_stats(): Enables creation of statistic plot animations visualizing possible relationships between movement paths and basemap. Define individual plot designs based on ggplot2 syntax.
 + stats arguments for animate_move(): Enables statistic plot animations side-by-side with the spatial plot animation of animate_move. Use the animate_stats() arguments with animate_move().
 + extent_factor argument for animate_move(): Increase the distance between the spatial extents of the movement paths and the basemap.

Improvements:
 + Several minor bug fixes


-----------------------------------------------------------
moveVis 0.9.3
Fixing major bug
-----------------------------------------------------------
Improvements
 + Fixed a major bug causing an unavoidable error when creating animations with animate_move() using user defined basemaps (Error message: 'Error in eval(expr, envir, enclos) : Object 'frame_l' not found')


-----------------------------------------------------------
moveVis 0.9.2
adding frame_width and frame_height arguments
-----------------------------------------------------------
New features:
 + Added two new arguments to animate_move() function for changing output frame resolution
 
Bug report contributors:
 + vestlink@github

 
-----------------------------------------------------------
moveVis 0.9.1
Fixing important bug; adding animate_raster()
-----------------------------------------------------------
New features:

R/move_data.R
#' Movement data of a White Stork population located nearby Lake Constance, Germany
#'
#' Dataset containing longitude/latitude point cooridnates and acquisition times of several white stork individuals
#' of a population located nearby Lake Constance.
#'
#' \itemize{
#'   \item lon. Longitude
#'   \item lat. Latitude
#'   \item individual. Name of the individual
#'   \item population. Name of the popualtion
#'   \item dt. Date and timestamps (to be used e. g. as POSIXct)
#'   ...
#' }
#' @details The example data have been pre-processed as explained in the examples. Please note that a very basic moving-/non-moving segmentation algorithm has been used in order to keep the code simple.
#' The provided code and data cannot be applied for actual data analysis, only adequate for movement data visualization! 
#' 
#' @format Data frame containing 2408 rows and 5 variables
#' @source Movebank (2013): URL \url{http://www.movebank.org/}
#' 
#' @examples
#' #DISCLAIMER: The provided code and data cannot be applied for actual data analysis
#' #and should only show the derivation of the example data provided with moveVis.
#' #This code is only adequate for movement data visualization.
#' 
#' \dontrun{
#' #Calculate start/stop times, FUNCTION
#' st_times <- function(data){
#'   for(i in 1:length(data)){
#'     if(i == 1){
#'       start_dt <- data[[i]]$dt[1]
#'       stop_dt <- data[[i]]$dt[length(data[[i]]$dt)]
#'     }else{
#'       start_dt <- c(start_dt, data[[i]]$dt[1])
#'       stop_dt <- c(stop_dt, data[[i]]$dt[length(data[[i]]$dt)])
#'     }
#'   }
#'   return(list(start_dt,stop_dt))
#' }

#' #Distance recalculation, FUNCTION
#' calc_dist <- function(data){
#'   p1 <- data[1:(length(data$lon)-1),1:2]
#'   p2 <- data[2:length(data$lon),1:2]
#'   data$dist_m <- NA
#'   data$dist_m[2:length(data$dist_m)] <- distGeo(p1,p2)
#'   return(data)
#' }
#' 
#' #Moving/non-moving segmentator, FUNCTION
#' class_moving <- function(data,resting_rad){
#'   data$moving <- 0
#'   data$moving[which(data$dist_m >= resting_rad)] <- 1
#'   return(data)
#' }
#' 
#' #Read data
#' data_dir <- "/path/to/movedata.txt" #In this case, an ASCII file with multiple populations
#' dr <- read.table(data_dir,header=TRUE,sep=";")
#' pop_levels <- levels(dr$population)
#' pop_levels_n <- length(pop_levels)
#' 
#' #Differentiate data per population
#' for(i in 1:pop_levels_n){
#'   if(i == 1){
#'     pop_subset <- list(subset(dr, population == pop_levels[i]))
#'   }else{
#'     pop_subset <- c(pop_subset,list(subset(dr, population == pop_levels[i])))
#'   }
#' }
#' 
#' #Differentiate data per individual
#' indi_levels <- levels(dr$individual)
#' indi_levels_n <- length(indi_levels)
#' for(i in 1:indi_levels_n){
#'   if(i == 1){
#'     indi_subset <- list(subset(dr, individual == indi_levels[i]))
#'   }else{
#'     indi_subset <- c(indi_subset,list(subset(dr, individual == indi_levels[i])))
#'   }
#' }
#' 
#' #Compute animation input list with all individuals per population
#' pop_select <- 2 #Selectin a population within the data
#' match_indi_subs <- as.integer(na.omit(match(indi_levels,pop_subset[[pop_select]]$individual)))
#' for(i in 1:length(match_indi_subs)){
#'   if(i == 1){
#'     indi_subset_pop <- list(pop_subset[[pop_select]][match_indi_subs[i]:
#'     (match_indi_subs[i+1]-1),])
#'   }else{
#'     if(i != length(match_indi_subs)){
#'       indi_subset_pop[i] <- list(pop_subset[[pop_select]][match_indi_subs[i]:
#'          (match_indi_subs[i+1]-1),])
#'     }else{
#'       indi_subset_pop[i] <- list(pop_subset[[pop_select]][match_indi_subs[i]:
#'          length(pop_subset[[pop_select]][,1]),])
#'     }
#'   }
#' }
#' 
#' #Calculate dt stamps, extract start_dt and stop_dt for all arrays
#' for(i in 1:length(indi_subset_pop)){
#'   dt <- c(paste0(indi_subset_pop[[i]]$date, " ",indi_subset_pop[[i]]$time))
#'   dt_stamps <- as.POSIXct(strptime(dt, "%Y-%m-%d %H:%M:%S", tz = "UTC"))
#'   indi_subset_pop[[i]]$dt <- align.time(dt_stamps, n=60)
#' }
#' start_dt <- st_times(indi_subset_pop)[[1]]
#' stop_dt <- st_times(indi_subset_pop)[[2]]
#' 
#' #Animal specifications, here for the White Stork
#' max_speed <- 45 
#' max_speed_min <- max_speed/60 #km/min
#' tolerance_m_min <- 50 #meter
#' temp_res <- 5 #for simplicity, we do not detect the temp. resolution automatically here
#' 
#' max_speed_spec <- max_speed_min*temp_res #km/temp_res
#' max_dist_m <- (max_speed_spec*1000)+tolerance_m_min #m/temp_res + tolerance range in m/temp_res
#' 
#' #Clean up individual data by defined props (eliminating peaks etc.), store them per individual
#' for(i in 1:length(indi_subset_pop)){
#'   data_clean <- indi_subset_pop[[i]]
#'   no_peaks <- FALSE
#'   while(no_peaks == FALSE){
#'     data_clean <- calc_dist(data_clean)
#'     dist_peaks <- which(is.na(data_clean$dist_m) == FALSE & data_clean$dist_m >= max_dist_m)
#'     if(length(dist_peaks) <= 1){
#'       no_peaks <- TRUE
#'     } else {
#'       data_clean <- data_clean[-dist_peaks[1:length(dist_peaks)],]
#'     }
#'   }
#'   if(i == 1){indi_subset_clean <- list(data_clean)
#'   }else{indi_subset_clean[i] <- list(data_clean)
#'   }
#'   indi_subset_clean[[i]]$peak <- 0
#'   indi_subset_clean[[i]]$peak[
#'      which(is.na(indi_subset_clean[[i]]$dist_m) == FALSE &
#'         indi_subset_clean[[i]]$dist_m >= max_dist_m)] <- 1 
#' }
#' 
#' #Recalculate start/stop times
#' start_dt <- st_times(indi_subset_clean)[[1]]
#' stop_dt <- st_times(indi_subset_clean)[[2]]
#' 
#' #Interpolate tracks
#' for(i in 1:length(indi_subset_clean)){
#'   out_n <- as.integer(difftime(stop_dt[i],start_dt[i],units="mins"))+1
#'   coords <- matrix(c(indi_subset_clean[[i]]$lon,indi_subset_clean[[i]]$lat),
#'      nrow=length(indi_subset_clean[[i]]$lon))
#'   lon_inter <- approx(coords[,1], n = out_n)
#'   lat_inter <- approx(coords[,2], n = out_n)
#'   dt_new <- seq(indi_subset_clean[[i]]$dt[1],
#'      indi_subset_clean[[i]]$dt[length(indi_subset_clean[[i]]$dt)], length.out = out_n)
#'   
#'   pop_name <- as.character(dt_new)
#'   pop_name[1:length(pop_name)] <- as.character(indi_subset_clean[[i]]$population[1])
#'   indi_name <- as.character(dt_new)
#'   indi_name[1:length(indi_name)] <- as.character(indi_subset_clean[[i]]$individual[1])
#'   if(i==1){
#'     indi_subset_int <- list(data.frame(lon_inter$y,lat_inter$y,pop_name,indi_name,dt_new))
#'     colnames(indi_subset_int[[i]]) <- c("lon","lat","population","individual","dt")
#'   }else{
#'     indi_subset_int[i] <- list(data.frame(lon_inter$y,lat_inter$y,pop_name,indi_name,dt_new))
#'     colnames(indi_subset_int[[i]]) <- c("lon","lat","population","individual","dt")
#'   }
#'   indi_subset_int[[i]] <- calc_dist(indi_subset_int[[i]])
#' }
#' 
#' #Segmentate moving/resting
#' resting_rad <- 20 #meters/min
#' for(i in 1:length(indi_subset_int)){
#'   if(i == 1){
#'     indi_subset_class <- list(class_moving(indi_subset_int[[i]],resting_rad))
#'   }else{
#'     indi_subset_class[i] <- list(class_moving(indi_subset_int[[i]],resting_rad))
#'   }
#' }
#' 
#' #Removing non-moving time periods...
#' for(i in 1:length(indi_subset_class)){
#'   if(i == 1){indi_subset_moving <- list(
#'      indi_subset_class[[i]][which(indi_subset_class[[i]]$moving == 1),])}
#'   else{indi_subset_moving[i] <- list(
#'      indi_subset_class[[i]][which(indi_subset_class[[i]]$moving == 1),])}
#' }
#' 
#' #Extract names of each individual
#' for(i in 1:length(indi_subset_moving)){
#'   if(length(indi_subset_moving[[i]][,1]) > 0){
#'     if(i==1){
#'       indi_names <- indi_subset_moving[[i]]$individual[1]
#'     }else{
#'       indi_names <- paste0(indi_names,", ",indi_subset_moving[[i]]$individual[1])
#'     }
#'   }
#' }
#' 
#' #Create move objects list
#' move_index <- 0
#' for(i in 1:length(indi_subset_moving)){
#'   maxi <- length(indi_subset_moving[[i]]$lon)
#'   #maxi <- 200 #If the maximum length should be defined by the user
#'   #if(length(indi_subset_moving[[i]]$lon) < maxi){
#'      maxi <- length(indi_subset_moving[[i]]$lon)
#'   }
#'   
#'   if(length(indi_subset_moving[[i]][,1]) > 0){
#'     if(i == 1){data_ani <- 
#'        list(move(x=indi_subset_moving[[i]]$lon[1:maxi],y=indi_subset_moving[[i]]$lat[1:maxi],
#'           time=indi_subset_moving[[i]]$dt[1:maxi],proj=CRS("+proj=longlat +ellps=WGS84"),
#'           animal=unlist(strsplit(indi_names,", "))[i]))
#'     }else{data_ani[i-move_index] <-
#'        list(move(x=indi_subset_moving[[i]]$lon[1:maxi],y=indi_subset_moving[[i]]$lat[1:maxi],
#'           time=indi_subset_moving[[i]]$dt[1:maxi],proj=CRS("+proj=longlat +ellps=WGS84"),
#'          animal=unlist(strsplit(indi_names,", "))[i]))}
#'   }else{move_index <- move_index+1}
#' }
#' 
#' #write out
#' file.create("moveVis_sample.txt")
#' for(i in 1:length(indi_subset_moving)){
#'   print(i)
#'   if(i == 1){
#'     writethis <- indi_subset_moving[[i]]
#'     write.table(writethis,file="samples.txt",sep=";")
#'   }else{
#'     writethis <- rbind(writethis,indi_subset_moving[[i]])
#'     write.table(writethis,file="samples.txt",sep=";")
#'   }
#' }
#' }
"move_data"